### PR TITLE
[feat]: adapt for rk3588, supports packaging guest VM image files at compiling

### PR DESCRIPTION
**Pull request by luodeb**
Adapting for rk3588.

Loading the OS Image
1. Add configuration items in configs/*.toml.
``` toml
...

# load from memory
image_location = "memory"
kernel_path = "nimbos-aarch64.bin"
# dtb_path = "nimbos-aarch64.dtb"

# load from file system
# image_location = "fs"
# kernel_path = "nimbos-aarch64.bin"

memory_regions = [
...
```

``` rust
// axvm TODO: ...
pub struct AxVMCrateConfig {
    /// The location of the image, default is 'fs'.
    pub image_location: Option<String>, 
}
```

Then we can use `make A=(pwd) LOG=debug ARCH=aarch64 BLK=y VM_CONFIGS=configs/nimbos-aarch64.toml run` to start VMs. 
use `make A=(pwd) LOG=debug ARCH=aarch64 BLK=y VM_CONFIGS=configs/rk3588-aarch64.toml kernel` to build RK3588 kernel.

### CODE REVIEW

File: arceos-vmm/configs/rk3588-aarch64.toml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# load from memory
+image_location = "memory" 
+kernel_path = "linux-rk3588-aarch64.bin"
+dtb_path = "linux-rk3588.dtb"
+
+# load from file system
+# image_location = "fs"
+# kernel_path = "arceos-aarch64.bin"
+
+platform = "aarch64-rk3588j"
```
Comment (hky1999) [inline arceos-vmm/configs/rk3588-aarch64.toml:21]: why did you add platform related infos in guest VM's config file...

File: arceos-vmm/tools/rk3588/boot.its
Comment (hky1999) [inline arceos-vmm/tools/rk3588/boot.its:1]: we'd better move these BSP files into a dependent repo

**Review by hky1999 — CHANGES_REQUESTED** [review]
Sorry for delay.
Two main questions:
1. move rk3588 related BSP files into a seperate repo, use scripts to get them.
2. where to put rk3588 platform specific config files, should we add a dir called `/platforms/` here?

File: Cargo.toml [hunk 2]
```diff
@@ -7,9 +7,10 @@ members = ["arceos-vmm"]
 lto = true
 
 [workspace.dependencies]
-axalloc = { git = "https://github.com/arceos-hypervisor/arceos.git", rev = "d938aca" }
-axhal = { git = "https://github.com/arceos-hypervisor/arceos.git", rev = "d938aca" }
-axstd = { git = "https://github.com/arceos-hypervisor/arceos.git", rev = "d938aca" }
+axalloc = { git = "https://github.com/arceos-hypervisor/arceos.git", rev = "3690f8e" }
```
[the remark this answers is not in the document]
Comment (luodeb) [reply Cargo.toml:10]: Sure

File: arceos-vmm/Makefile [hunk 3]
```diff
@@ -81,10 +81,12 @@ else ifneq ($(PLATFORM),)
   ifneq ($(filter $(PLATFORM),$(builtin_platforms)),)
     # builtin platform
     PLATFORM_NAME := $(PLATFORM)
+    VM_CONFIGS ?= configs/$(PLATFORM).toml
```
[the remark this answers is not in the document]
Comment (luodeb) [reply arceos-vmm/Makefile:84]: If want to load a VM image from memory, need to parse the toml. The VM_CONFIGS environment variable is used to specify the toml.

File: arceos-vmm/scripts/make/rk3588.mk [hunk 4]
```diff
@@ -0,0 +1,3 @@
+kernel: build
+	./tools/rk3588/mkimg --dtb rk3588-firefly-itx-3588j.dtb --img $(OUT_BIN)
```
[the remark this answers is not in the document]
Comment (luodeb) [reply arceos-vmm/scripts/make/rk3588.mk:2]: OK

File: arceos-vmm/src/vmm/config.rs [hunk 5]
```diff
@@ -2,7 +2,9 @@ use axvm::config::{AxVMConfig, AxVMCrateConfig};
 
 use crate::vmm::{images::load_vm_images, vm_list::push_vm, VM};
 
-mod config {
+core::arch::global_asm!(include_str!("../../guest.S"));
```
[the remark this answers is not in the document]
Comment (luodeb) [reply arceos-vmm/src/vmm/config.rs:5]: my bad.

File: arceos-vmm/configs/rk3588-aarch64.toml (hunk 1 above)
Comment (luodeb) [reply arceos-vmm/configs/rk3588-aarch64.toml:21]: The Makefile will read platform
image

File: arceos-vmm/tools/rk3588/boot.its
Comment (luodeb) [reply arceos-vmm/tools/rk3588/boot.its:1]: OK

File: arceos-vmm/configs/rk3588-aarch64.toml (hunk 1 above)
Comment (hky1999) [reply arceos-vmm/configs/rk3588-aarch64.toml:21]: Yep, but this is the platform config for ArceOS, instead of guest VM's config file, 
we should seperate them

File: arceos-vmm/build.rs [hunk 6]
```diff
@@ -111,5 +188,9 @@ fn main() -> io::Result<()> {
         }
     }
     writeln!(output_file, "}}")?;
+
+    // load kernel and dtb images
+    load_guest_img(output_file, config_toml_path)?;
```
[the remark this answers is not in the document]
Comment (luodeb) [reply arceos-vmm/build.rs]: OK

File: arceos-vmm/build.rs [hunk 7]
```diff
@@ -75,11 +77,118 @@ fn open_output_file() -> fs::File {
...
+}}
+"#
+    )?;
+    Ok(())
+}
+
 fn main() -> io::Result<()> {
+    let platform = env::var("AX_PLATFORM").unwrap_or("".to_string());
+    let platform_family = env::var("AX_PLATFORM").unwrap_or("".to_string());
+    let config_toml_path = get_config_paths();
```
Comment (hky1999) [inline arceos-vmm/build.rs:188]: I think this function is duplicated with the already existed `get_configs` function right?

File: arceos-vmm/build.rs [hunk 8]
```diff
@@ -75,11 +77,118 @@ fn open_output_file() -> fs::File {
...
+        if path.is_relative() {
+            fs::canonicalize(&configs_path).unwrap_or_else(|_| path.to_path_buf())
+        } else {
+            path.to_path_buf()
+        }
+    }
+
+    if let Some(config_path) = config_toml_path {
+        // Started from the first config item by default.
+        if let Some(guest_config) = config_path.first() {
```
Comment (hky1999) [inline arceos-vmm/build.rs:113]: do you consider if multiple VM config files are privided, and we have to generate multiple `load_xxx` functions for each VM.

It seems that you only load the first one.

Comment (hky1999) [reply arceos-vmm/build.rs:113]: we can use a array here, with `vm_id` as index?

something like `get_kernel_binaries_of_vm(index: usize) -> Option<&'static [u8]>`

File: arceos-vmm/build.rs [hunk 9]
```diff
@@ -75,11 +77,118 @@ fn open_output_file() -> fs::File {
...
+                if location == "memory" {
+                    let kernel_path = convert_to_absolute(
+                        "configs",
+                        config.get("kernel_path").unwrap().as_str().unwrap(),
+                    );
+
+                    // If have dtb_path, include it.
+                    writeln!(
+                        out_file,
+                        r#"pub fn load_dtb() -> Option<&'static [u8]> {{ "#
```
Comment (hky1999) [inline arceos-vmm/build.rs:128]: load_xxx is not a good function name

use something like get_xxx_binaries

File: arceos-vmm/src/vmm/images.rs [hunk 10]
```diff
@@ -9,9 +9,21 @@ use axvm::config::AxVMCrateConfig;
...
+    match config.image_location.as_deref() {
+        Some("memory") => load_vm_images_memory(config, vm),
+        _ => load_vm_images_filesystem(config, vm),
+    }
+}
+
 /// Loads the VM image files from the filesystem
 /// into the guest VM's memory space based on the VM configuration.
-pub fn load_vm_images(config: AxVMCrateConfig, vm: VMRef) -> AxResult {
+fn load_vm_images_filesystem(config: AxVMCrateConfig, vm: VMRef) -> AxResult {
```
Comment (hky1999) [inline arceos-vmm/src/vmm/images.rs:25]: load_vm_images_from_filesystem

File: arceos-vmm/src/vmm/images.rs [hunk 11]
```diff
@@ -93,3 +105,83 @@ fn open_image_file(file_name: &str) -> AxResult<(File, usize)> {
         .size() as usize;
     Ok((file, file_size))
 }
+
+/// Load VM images from memory (guest.S)
+/// into the guest VM's memory space based on the VM configuration.
+fn load_vm_images_memory(config: AxVMCrateConfig, vm: VMRef) -> AxResult {
```
Comment (hky1999) [inline arceos-vmm/src/vmm/images.rs:111]: load_vm_images_from_memory

File: arceos-vmm/src/vmm/images.rs [hunk 12]
```diff
@@ -93,3 +105,83 @@ fn open_image_file(file_name: &str) -> AxResult<(File, usize)> {
         .size() as usize;
     Ok((file, file_size))
 }
+
+/// Load VM images from memory (guest.S)
```
Comment (hky1999) [inline arceos-vmm/src/vmm/images.rs:109]: guest.S?

update all comments please

File: arceos-vmm/src/vmm/images.rs [hunk 13]
```diff
@@ -93,3 +105,83 @@ fn open_image_file(file_name: &str) -> AxResult<(File, usize)> {
...
+        )
+        .expect("Failed to load VM images");
+    } else {
+        panic!("VM images is missed, add `VM_CONFIGS=configs/aarch64-linux.toml` in make command.");
+    }
+
+    Ok(())
+}
+
+fn load_vm_image_memory(
```
Comment (hky1999) [inline arceos-vmm/src/vmm/images.rs:152]: from_memory

File: arceos-vmm/src/vmm/images.rs [hunk 14]
```diff
@@ -93,3 +105,83 @@ fn open_image_file(file_name: &str) -> AxResult<(File, usize)> {
...
+    let mut buffer_pos = 0;
+    let image_load_gpa = GuestPhysAddr::from(load_addr);
+
+    let image_load_regions = vm.get_image_load_region(image_load_gpa, image_size)?;
+
+    for region in image_load_regions {
+        let region_len = region.len();
+        let bytes_to_write = region_len.min(image_size - buffer_pos);
+
+        // copy data from .tbdata section
```
Comment (hky1999) [inline arceos-vmm/src/vmm/images.rs:167]: from .tbdata section?

**Review by hky1999 — CHANGES_REQUESTED** [review]

File: arceos-vmm/configs/rk3588-aarch64.toml
Comment (hky1999) [inline arceos-vmm/configs/rk3588-aarch64.toml:1]: linux-rk3588-aarch64

File: arceos-vmm/platforms/aarch64-rk3588j.toml
Comment (hky1999) [inline arceos-vmm/platforms/aarch64-rk3588j.toml:1]: I think there are some design flaws at `Makefile` line 113 about variable `PLATFORM` https://github.com/arceos-hypervisor/arceos-umhv/blob/7dd3a337524651019586182b8d30be27a1f0226a/arceos-vmm/Makefile#L113

currently we fix `PLATFORM = platforms/aarch64-qemu-virt-hv.toml`, which is a bad design.

we can think about how to resolve that together

**Review by hky1999 — CHANGES_REQUESTED** [review]

File: arceos-vmm/build.rs [hunk 15]
```diff
@@ -75,11 +77,118 @@ fn open_output_file() -> fs::File {
...
+        let configs_path = Path::new(configs_path).join(path);
+        if path.is_relative() {
+            fs::canonicalize(&configs_path).unwrap_or_else(|_| path.to_path_buf())
+        } else {
+            path.to_path_buf()
+        }
+    }
+
+    if let Some(config_path) = config_toml_path {
+        // Started from the first config item by default.
```
Comment (hky1999) [inline arceos-vmm/build.rs:112]: enrich your comments of why do you select the first config item

File: arceos-vmm/build.rs [hunk 16]
```diff
@@ -75,11 +77,118 @@ fn open_output_file() -> fs::File {
...
+                if location == "memory" {
+                    let kernel_path = convert_to_absolute(
+                        "configs",
+                        config.get("kernel_path").unwrap().as_str().unwrap(),
+                    );
+
+                    // If have dtb_path, include it.
+                    writeln!(
+                        out_file,
+                        r#"pub fn get_dtb_binaries() -> Option<&'static [u8]> {{ "#
```
Comment (hky1999) [inline arceos-vmm/build.rs:128]: get_dtb_binary
since you only return one image file

File: arceos-vmm/build.rs [hunk 17]
```diff
@@ -75,11 +77,118 @@ fn open_output_file() -> fs::File {
...
+                        // use include_bytes! load image
+                        writeln!(out_file, r#"    Some(include_bytes!({:?}))"#, dtb_path)?;
+                    } else {
+                        writeln!(out_file, r#"    None"#)?;
+                    };
+
+                    // If have bios_path, include it.
+                    writeln!(
+                        out_file,
+                        r#"pub fn get_bios_binaries() -> Option<&'static [u8]> {{ "#
```
Comment (hky1999) [inline arceos-vmm/build.rs:141]: get_bios_binary

File: arceos-vmm/build.rs [hunk 18]
```diff
@@ -75,11 +77,118 @@ fn open_output_file() -> fs::File {
...
+                        writeln!(out_file, r#"    Some(include_bytes!({:?}))"#, bios_path)?;
+                    } else {
+                        writeln!(out_file, r#"    None"#)?;
+                    };
+
+                    writeln!(out_file, r#"}}"#)?;
+
+                    writeln!(
+                        out_file,
+                        r#"pub fn get_kernel_binaries() -> Option<&'static [u8]> {{ "#
```
Comment (hky1999) [inline arceos-vmm/build.rs:155]: get_kernel_binary

File: arceos-vmm/build.rs [hunk 19]
```diff
@@ -75,11 +77,118 @@ fn open_output_file() -> fs::File {
...
+
+                    return Ok(());
+                }
+            }
+        }
+    }
+    writeln!(
+        out_file,
+        r#"
+pub fn get_dtb_binaries() -> Option<&'static [u8]> {{
```
Comment (hky1999) [inline arceos-vmm/build.rs:169]: binary

File: arceos-vmm/build.rs [hunk 20]
```diff
@@ -75,11 +77,118 @@ fn open_output_file() -> fs::File {
...
+        }
+    }
+    writeln!(
+        out_file,
+        r#"
+pub fn get_dtb_binaries() -> Option<&'static [u8]> {{ 
+    None
+}}
+    
+pub fn get_kernel_binaries() -> Option<&'static [u8]> {{
```
Comment (hky1999) [inline arceos-vmm/build.rs:173]: binary

File: arceos-vmm/build.rs [hunk 21]
```diff
@@ -75,11 +77,118 @@ fn open_output_file() -> fs::File {
...
+        r#"
+pub fn get_dtb_binaries() -> Option<&'static [u8]> {{ 
+    None
+}}
+    
+pub fn get_kernel_binaries() -> Option<&'static [u8]> {{ 
+    None
+}}
+
+pub fn get_bios_binaries() -> Option<&'static [u8]> {{
```
Comment (hky1999) [inline arceos-vmm/build.rs:177]: binary

File: arceos-vmm/build.rs (hunk 7 above)
Comment (hky1999) [reply arceos-vmm/build.rs:188]: resolve this please

File: arceos-vmm/build.rs [hunk 22]
```diff
@@ -111,5 +221,9 @@ fn main() -> io::Result<()> {
         }
     }
     writeln!(output_file, "}}")?;
+
+    // generate "load kernel and dtb images function"
+    generate_load_guest_img(output_file, config_toml_path)?;
```
Comment (hky1999) [inline arceos-vmm/build.rs:226]: generate_guest_img_loading_functions

File: arceos-vmm/configs/nimbos-aarch64.toml [hunk 23]
```diff
@@ -4,10 +4,20 @@ vm_type = 1
 cpu_num = 1
 phys_cpu_sets = [1]
 entry_point = 0x4008_0000
-# bios_path = ""
-# bios_load_addr = 0
+
+# The location of image: "memory" | "fs"
+# load from memory
+# image_location = "memory"
+# kernel_path = "../../../guest/nimbos/kernel/target/aarch64/release/nimbos.bin"
```
Comment (hky1999) [inline arceos-vmm/configs/nimbos-aarch64.toml:11]: what are these configs for

File: README.md [hunk 24]
```diff
@@ -33,11 +33,14 @@ WIP on Linux guests.
...
 $ make disk_img
+$
+$ # Copy guest VM binary image files.
 $ mkdir -p tmp
 $ sudo mount disk.img tmp
-$ # Copy guest VM binary image files.
 $ sudo cp /PATH/TO/YOUR/GUEST/VM/IMAGE tmp/
 $ sudo umount tmp
+$
+$ # Otherwise, set `image_location = "memory"` in CONFIG/FILE, then set kernel_path
```
Comment (hky1999) [inline README.md:43]: add more descriptions here.

you can create a `doc` directory in the root path of this repo and privade guides about how to boot ArceOS-hypervisor upon rk3588 platform, as well as how to set Linux as its first VM.

**Review by hky1999 — APPROVED** [review]
Please fix these review comments.
thanks for your hard work

also, remember to catch up with the lates updates from master branch

File: arceos-vmm/build.rs [hunk 25]
```diff
@@ -79,7 +81,117 @@ fn open_output_file() -> fs::File {
...
+
+                    return Ok(());
+                }
+            }
+        }
+    }
+    writeln!(
+        out_file,
+        r#"
+pub fn get_dtb_binary() -> Option<&'static [u8]> {{
```
Comment (hky1999) [inline arceos-vmm/build.rs:172]: Who provides DTB for linux

**Review by hky1999 — DISMISSED** [review]

**Review by hky1999 — APPROVED** [review]
